flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -290,17 +290,33 @@ private void registerCallback() {
     @Override
     public void write(InputT element, Context context) throws IOException, InterruptedException {
         while (bufferedRequestEntries.size() >= maxBufferedRequests) {
-            mailboxExecutor.tryYield();
+            flush();
         }
 
         addEntryToBuffer(elementConverter.apply(element, context), false);
 
-        flushIfAble();
+        nonBlockingFlush();
     }
 
-    private void flushIfAble() {
-        while (bufferedRequestEntries.size() >= maxBatchSize
-                || bufferedRequestEntriesTotalSizeInBytes >= maxBatchSizeInBytes) {
+    /**
+     * Determines if a call to flush will be non-blocking (i.e. {@code inFlightRequestsCount} is
+     * strictly smaller than {@code maxInFlightRequests}). Also requires one of the following
+     * requirements to be met:
+     *
+     * <ul>
+     *   <li>The number of elements buffered is greater than or equal to the {@code maxBatchSize}
+     *   <li>The sum of the size in bytes of all records in the buffer is greater than or equal to
+     *       {@code maxBatchSizeInBytes}
+     * </ul>
+     */
+    private void nonBlockingFlush() {
+        boolean uncompletedInFlightResponses = true;
+        while (uncompletedInFlightResponses) {
+            uncompletedInFlightResponses = mailboxExecutor.tryYield();
+        }
+        while (inFlightRequestsCount < maxInFlightRequests
+                && (bufferedRequestEntries.size() >= maxBatchSize
+                        || bufferedRequestEntriesTotalSizeInBytes >= maxBatchSizeInBytes)) {
```

Comment:
3. I have added a condition here to only call `flush()` if we know beforehand it is not blocking. This is desirable because the element to be written has already been added to the buffer, so should not proceed to flush unless we are certain that operation will not block.

Reply:
4. This raises an interesting point. We know flush() will block if and only if there are too many in flight requests (`inFlightRequestsCount >= maxInFlightRequests`) AND if at least one of those in flight requests have not yet called the `completeRequest` method. 

5. If we determine whether at least one of those in flight requests have not yet called the `completeRequest` method, then we can accurately ascertain whether a call to flush will block or not.

Reply:
The resulting behaviour is: "If we have enough elements in the buffer to make a batch and we can flush that without blocking, then we should do that in the earliest possible opportunity"